com.unity.shadergraph/Editor/Data/Nodes/AbstractMaterialNode.cs

```diff
@@ -756,6 +756,11 @@ protected virtual void CalculateNodeHasError()
             }
         }
 
+        protected string GetRayTracingError() => $@"
+            #if defined(SHADER_STAGE_RAY_TRACING)
+            #error '{name}' node is not supported in ray tracing, please provide an alternate implementation, relying for instance on the 'Raytracing Quality' keyword
+            #endif";
```

Comment:
I'll have our designer take a look, but I suspect we will want to implement this in an entirely different way.

Reply:
Since we can currently declare incompatibility at the target level for nodes on a graph at editor time, I'd prefer to look into managing these errors in the same way.  Take a look at what the user sees when they use an HDRP specific node in a shader graph with a URP target active, for reference of what we're wanting here.

Reply:
Hmm..  the tricky bit is that these are only really errors in very specific circumstances -- when the ShaderGraph is actually used in ray tracing.  If the ShaderGraph is used for other things it shouldn't be an error.  So it's unlike the existing render pipeline checks.

There's no "Enable Raytracing" toggle on ShaderGraph (as far as I can tell).. so no indication if it's intended to be used with ray tracing.  Do we just check the global raytracing enable setting?

How do we know the ShaderGraph isn't some other shader that isn't being used for ray tracing?  Like a VFX shadergraph or something that's not rendered via ray tracing?

We could make a "ray tracing compatibility" view mode of some kind, that would show the warnings on the nodes that are not compatible.

Thoughts?

Reply:
Hmm, I guess we could also modify the Target.IsNodeAllowedByTarget() to provide graph-level warnings only when RayTracing is enabled globally in the project.  Would that have any issues?

Reply:
i.e. flag this nodes with [UsesDerivativeInstructions] or something, then have Target.IsNodeAllowedByTarget() react to that

Reply:
We might need some way for the user to indicate at the Target level that it's ok to use derivatives in this graph, even though raytracing is enabled, because this graph is not intended to be ray traced...

Reply:
The error is only triggered when compiled by active ray tracing shader, the user will never see anything unless he enables ray tracing and uses this shader to render ray tracing effects

Reply:
The thing is we son't want to prevent users to use in the rasterization version of the shader, but we want to warn them that it is not okay for ray tracing
so flagging the whole shader graph or even the node "ray tracing" compatible will not work

Reply:
Question is: how do we give the user feedback inside the ShaderGraph editor about which nodes are not ray-tracing compatible, if they do want to use it for ray tracing?

Reply:
> It would be nice to be able to do that ^ if we can.

I agree, and I actually tried to go for that, but as it's not possible yet (and requires work way outside the scope of this PR), it'd be great if we could land it as is, since it is already a usability improvement over the very cryptic, global errors we get atm when such nodes are present in a graph, when activating ray tracing.

If in the future we can replace/complement that with visual hints during graph authoring, that'd be even better of course!